syntax = "proto3";


package user;

message User {
  int32 id = 1;
  string username = 2;
  string email = 3;
}

message CreateUserRequest {
  string username = 1;
  string email = 2;
}

message GetUserRequest {
  int32 id = 1;
}

message UpdateUserRequest {
  int32 id = 1;
  string username = 2;
  string email = 3;
}

message DeleteUserRequest {
  int32 id = 1;
}

message ListUsersRequest {}

message UserResponse {
  User user = 1;
}

message ListUsersResponse {
  repeated User users = 1;
}

// 根据用户名称 模糊搜索
message SearchUsersRequest {
  string query = 1;
}
message SearchUsersResponse {
  repeated User users = 1;
}

// 只根据邮箱搜索
message SearchUsersByEmailRequest {
  string email = 1;
}
message SearchUsersByEmailRespose{
  // 精确匹配 返回单个用户对象
  User user = 1;
}


service UserService {
  rpc CreateUser(CreateUserRequest) returns (UserResponse);
  rpc GetUser(GetUserRequest) returns (UserResponse);
  rpc UpdateUser(UpdateUserRequest) returns (UserResponse);
  rpc DeleteUser(DeleteUserRequest) returns (UserResponse);
  rpc ListUsers(ListUsersRequest) returns (ListUsersResponse);
  rpc SearchUsers(SearchUsersRequest) returns (SearchUsersResponse);
}
